Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift

```diff
@@ -5,27 +5,106 @@ import MapboxCoreMaps
     var northwest: CLLocationCoordinate2D {
-        return CLLocationCoordinate2D(latitude: northeast.latitude, longitude: southwest.longitude)
+        return __northwest()
     }
 
-    /// The northwest coordinate of an internal `CoordinateBounds` type.
+    /// The southeast coordinate of the bounds.
     var southeast: CLLocationCoordinate2D {
-        return CLLocationCoordinate2D(latitude: southwest.latitude, longitude: northeast.longitude)
+        return __southeast()
     }
-}
 
-internal extension CoordinateBounds {
-    func contains(_ coordinates: [CLLocationCoordinate2D]) -> Bool {
-        let latitudeRange = southwest.latitude...northeast.latitude
-        let longitudeRange = southwest.longitude...northeast.longitude
+    /// Returns a new `CoordinateBounds` that stretches to contain both this and another `CoordinateBounds`.
+    /// - Parameter coordinateBounds: The `CoordinateBounds` to add.
+    /// - Returns: A bounds whose area encompasses the two bounds and the space between them.
+    func union(_ coordinateBounds: CoordinateBounds) -> CoordinateBounds {
+        extend(forArea: coordinateBounds)
+    }
 
-        for coordinate in coordinates {
-            if latitudeRange.contains(coordinate.latitude) || longitudeRange.contains(coordinate.longitude) {
-                return true
-            }
+    /// Extends these bounds to include `point`.
+    /// - Parameter coordinate: The coordinate to be included.
+    /// - Returns: The extended bounds.
+    func include(_ coordinate: CLLocationCoordinate2D) -> CoordinateBounds {
+        extend(forPoint: coordinate)
+    }
+
+    /// Returns a new `CoordinateBounds` that is the intersection of this with another `CoordinateBounds`.
+    /// - Parameter coordinateBounds: The `CoordinateBounds` to intersect with.
+    /// - Returns: The bounds representing the intersection of the two bounds or null if there is no intersection.
+    func intersect(_ coordinateBounds: CoordinateBounds) -> CoordinateBounds? {
+        let minWest = max(coordinateBounds.west, west)
+        let maxEast = min(coordinateBounds.east, east)
+
+        guard maxEast >= minWest else { return nil }
+
+        let minSouth = max(coordinateBounds.south, south)
+        let maxNorth = min(coordinateBounds.north, north)
+
+        guard maxNorth >= minSouth else { return nil }
+
+        return CoordinateBounds(southwest: CLLocationCoordinate2D(latitude: minSouth, longitude: minWest),
+                                northeast: CLLocationCoordinate2D(latitude: maxNorth, longitude: maxEast))
+    }
+
+    // MARK: - Equality
+
+    override var hash: Int {
+        return Int((north + 90)
+        + (south + 90) * 1000
+        + (east + 180) * 1000000
+        + (west + 180) * 1000000000)
+    }
+
+    override func isEqual(_ object: Any?) -> Bool {
+        guard let other = object as? CoordinateBounds else {
+            return false
         }
-        return false
+
+        return self == other
+    }
```

Comment:
I'm wary of overriding methods in extensions. Is it safe?

Reply:
Seems it's allowed only for extensions of Objective-C types. What's the correct mental model for how it works? What happens if the same type has its own implementation?

Reply:
Not that safe it seems:

> If the name of a method declared in a category is the same as a method in the original class, or a method in another category on the same class (or even a superclass), the behavior is undefined as to which method implementation is used at runtime. This is less likely to be an issue if you’re using categories with your own classes, but can cause problems when using categories to add methods to standard Cocoa or Cocoa Touch classes.

From [here](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html), removed them.